server/models/annotation.py

```diff
@@ -727,3 +776,109 @@ def validate(self, doc):
+                continue
+            _versions.add(entry['_version'])
+            # test the itemId versus the user
+            if not isAdmin:
+                try:
+                    self.requireAccess(entry, user, AccessType.READ)
+                except AccessException:
+                    continue
+            if offset:
+                offset -= 1
+                continue
+            yield entry
+            if limit:
+                limit -= 1
+                if not limit:
+                    break
+
+    def getVersion(self, annotationId, version, user=None, force=False, *args, **kwargs):
+        """
+        Get an annotation history version.  This reconstructs the original
+        annotation.
+
+        :param annotationId: the annotation to get history for.
+        :param version: the specific version to get.
+        :param user: the Girder user.  If the user is not an admin, they must
+            have read access on the item and the item must exist.
+        :param force: if True, don't get the user access.
+        """
+        if annotationId and not isinstance(annotationId, ObjectId):
+            annotationId = ObjectId(annotationId)
+        entry = self.findOne({
+            '$or': [{'_id': annotationId}, {'_annotationId': annotationId}],
+            '_version': int(version)
+        })
+        if not entry:
+            return None
+        result = self.load(entry['_id'], user=user, force=force, *args, **kwargs)
+        result['_versionId'] = result['_id']
+        result['_id'] = result.pop('annotationId', result['_id'])
+        return result
+
+    def revertVersion(self, id, version=None, user=None, force=False):
+        """
+        Revert to a previous version of an annotation.
+
+        :param id: the annotation id.
+        :param version: the version to revert to.  None reverts to the previous
+            version.  If the annotation was deleted, this is the most recent
+            version.
+        :param user: the user doing the reversion.
+        :param force: if True don't authenticate the user with the associated
+            item access.
+        """
+        if version is None:
+            oldVersions = list(Annotation().versionList(id, limit=2, force=True))
+            if len(oldVersions) >= 1 and oldVersions[0].get('_active') is False:
+                version = oldVersions[0]['_version']
+            elif len(oldVersions) >= 2:
+                version = oldVersions[1]['_version']
+        annotation = Annotation().getVersion(id, version, user, force=force)
+        if annotation is None:
+            return
+        # If this is the most recent (active) annotation, don't do anything.
+        # Otherwise, revent it.
+        if not annotation.get('_active', True):
+            item = Item().load(annotation.get('itemId'), force=True)
+            if item is not None and not force:
+                Item().requireAccess(
+                    item, user=user, level=AccessType.WRITE)
```

Comment:
The access control is a thorny issue here.  I don't really like relying on the item access control here and the annotation access control elsewhere.  Even though it isn't ideal, I think using the access list of the current version of the annotation would be more consistent.  Perhaps we should avoid considering the access list part of the versioning.

Reply:
It should be sufficient to just validate that we can write the annotation.  I'll change this to:
```
if not force:
    self.requireAccess(annotation, user=user, level=AccessType.WRITE)
```